execution_engine/src/core/engine_state/mod.rs

```diff
@@ -1459,17 +1461,27 @@ where
             // session_code_spec_1: gas limit = ((balance of handle payment payment purse) /
             // gas_price)
             // - (gas spent during payment execution)
-            let session_gas_limit: Gas =
-                match Gas::from_motes(payment_purse_balance, deploy_item.gas_price)
-                    .and_then(|gas| gas.checked_sub(payment_result_cost))
-                {
-                    Some(gas) => gas,
-                    None => {
-                        return Ok(ExecutionResult::precondition_failure(
-                            Error::GasConversionOverflow,
-                        ))
-                    }
-                };
+            let gas_price = deploy_item.gas_price;
+
+            // Original expression from `session_code_spec_1` will fail if user really wants to pay
+            // more than a `Gas` counter can contain i.e. `Gas::MAX + 1`.
+            // By rewriting this expression: (payment_purse_balance/gas_price)-payment_result_cost
+            // into: (payment_purse_balance-gas_price*payment_result_cost) / gas_price
+            // we will not run into a gas overflow as we will first subtract payment result (in
+            // motes) and later we will convert the result into a gas.
+            let session_gas_limit: Gas = match Motes::from_gas(payment_result_cost, gas_price)
+                .and_then(|payment_result_motes| {
+                    payment_purse_balance.checked_sub(payment_result_motes)
+                })
+                .and_then(|motes| Gas::from_motes(motes, gas_price))
+            {
+                Some(gas) => gas,
+                None => {
+                    return Ok(ExecutionResult::precondition_failure(
+                        Error::GasConversionOverflow,
```

Comment:
It's worth mentioning we validate the account's main_purse to possess at least the `payment_result_cost` in https://github.com/casper-network/casper-node/blob/b84d3909d7d05d8b3865dff692dbbb9ea86f1c11/execution_engine/src/core/engine_state/execution_result.rs#L252-L258.

I believe by rewriting this expression to avoid overflow if the user pays Gas::MAX+1, I'm not changing the existing behavior